modules/auxiliary/scanner/scada/pcomclient.rb

```diff
@@ -0,0 +1,197 @@
+      print_error("Invalid ADDRESS")
+      return
+    end
+
+    case datastore['OPERAND']
+    when "Input"
+      cc = "RE"
+    when "Output"
+      cc = "RA"
+    when "SB"
+      cc = "GS"
+    when "MB"
+      cc = "RB"
+    when "MI"
+      cc = "RW"
+    when "SI"
+      cc = "GF"
+    when "ML"
+      cc = "RNL"
+    when "SL"
+      cc = "RNH"
+    else
+      print_error("Unknown operand #{datastore['OPERAND']}")
+      return
+    end
+
+    address = datastore['ADDRESS'].to_s(16).rjust(4,'0')
+    length = datastore['LENGTH'].to_s(16).rjust(2,'0')
+    print_status("Reading #{length} values (#{datastore['OPERAND']}) starting from #{address} address")
+    sock.put(pcom_ascii_request(cc + address + length))
+    sock.get_once
+  end
+
+  def print_read_ans(ans)
+    cc = ans[0..1]
+    data = ans[2..ans.length]
+    start_addr = datastore['ADDRESS']
+    case cc
+    when "RE"
+      size = 1
+    when "RA"
+      size = 1
+    when "RB"
+      size = 1
+    when "GS"
+      size = 1
+    when "RW"
+      size = 4
+    when "GF"
+      size = 4
+    when "RN"
+      size = 8
+    else
+      print_error("Unknown answer #{cc}")
+      return
+    end
+    data.scan(/.{#{size}}/).each_with_index {|val, i|
+      print_good("[#{(start_addr + i).to_s.rjust(5,'0')}] : #{val.to_i(16)}")}
+  end
+
+  def write
+    values = datastore['VALUES'].split(",")
+    case datastore['OPERAND']
+    when "Input"
+      print_error("Input operand is read only")
+      return
+    when "Output"
+      cc = "SA"
+    when "SB"
+      cc = "SS"
+    when "MB"
+      cc = "SB"
+    when "MI"
+      cc = "SW"
+    when "SI"
+      cc = "SF"
+    when "ML"
+      cc = "SNL"
+    when "SL"
+      cc = "SNH"
+    else
+      print_error("Unknown operand #{datastore['OPERAND']}")
+      return
+    end
+
+    address = datastore['ADDRESS'].to_s(16).rjust(4,'0')
+    length = values.length.to_s(16).rjust(2,'0')
+    values_to_write = values.map{|s| s.to_i(10).to_s(16).rjust(4,'0')}.join
+    print_status("Writing #{length} #{datastore['OPERAND']} (#{datastore['VALUES']}) starting from #{address} address")
+    sock.put(pcom_ascii_request(cc + address + length + values_to_write))
+    sock.get_once
+  end
+
+  def run
+    connect
+    case action.name
+    when "READ"
+      if datastore['LENGTH'] == nil
+        print_error("The option VALUES is not set")
+        return
+      else
+        ans = read
+        if ans == nil
+          print_error("No answer from PLC")
+          return
+        end
+        print_read_ans(ans.to_s[10..(ans.length-4)])
+      end
+    when "WRITE"
+      if datastore['VALUES'] == nil
+        print_error("The option VALUES is not set")
+        return
+      else
+        ans = write
+        if ans == nil
+          print_error("No answer from PLC")
+          return
+        end
+      end
+    else
+      print_error("Unknown action #{action.name}")
+    end
+    disconnect
```

Comment:
Good practice on the cleanup with `disconnect`! However this is actually automatically taken care of in the `Msf::Exploit::Remote::Tcp` minx. So you can remove this line safely and just let the mixin do its thing :-)

Reply:
done